### PR TITLE
Streams update 2.4.18

### PR DESCRIPTION
**Pull request by Arkatufus**
Port of https://github.com/akka/akka/pull/22657, https://github.com/akka/akka/pull/22647 and https://github.com/akka/akka/pull/22033.

Re-base from https://github.com/akkadotnet/akka.net/pull/2710/commits/59af0c2018446c98e18f302b3587c2c632b46e3a for LazySource class.

- https://github.com/akka/akka/pull/22657 port needs review.
- https://github.com/akka/akka/pull/22647 port needs review.
- https://github.com/akka/akka/pull/22033 port needs review.

### CODE REVIEW

**Comment by Silv3rcircl3** [comment]
@Arkatufus you need to update the ApiApproval file because you've changed the public API. 
To do so you need to run the Akka.API.Tests locally, after that you should see a diff window in VS, on the left the current approval file and on the right the `CoreAPISpec.ApproveStreams.received.txt`  and then simply copy everything from `CoreAPISpec.ApproveStreams.received.txt` into the left window, save and commit the `CoreAPISpec.ApproveStreams.approved.txt`

**Comment by Arkatufus** [comment]
Done :)

File: src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs [hunk 1]
```diff
@@ -209,6 +248,29 @@ public void SynchronousFileSink_should_allow_overriding_the_dispatcher_using_Att
             }, _materializer);
         }
 
+        // Needed help converting this test case
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs]: can this be removed ? :)

File: src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs [hunk 2]
```diff
@@ -209,6 +248,29 @@ public void SynchronousFileSink_should_allow_overriding_the_dispatcher_using_Att
...
+        public void SynchronousFileSink_should_write_single_line_to_a_file_from_lazy_sink()
+        {
+            this.AssertAllStagesStopped(() => {
+                TargetFile(f => {
+                    var lazySink = Sink.LazySink(
+                        (ByteString _) => Task.FromResult(FileIO.ToFile(f)),
+                        () => Task.FromResult(IOResult.Success(0)))
+                        .MapMaterializedValue(t => {
+                            t.Wait(TimeSpan.FromSeconds(3));
+                            return t.Result;
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs]: I think it's more like `t.ContinueWIth(r=>r.Result.Result)` but not sure if it matters, if the test passes leave it as it is ;)

File: src/core/Akka.Streams/Implementation/Sinks.cs [hunk 3]
```diff
@@ -974,8 +987,21 @@ private void Failure(Exception ex)
...
-                    .RunWith(sink, Interpreter.SubFusingMaterializer));
+
+                try {
+                    var matVal = Source.FromGraph(sourceOut.Source)
+                        .RunWith(sink, Interpreter.SubFusingMaterializer);
+                    _completion.TrySetResult(matVal);
+                } catch (Exception ex) {
+                    /*
+                     case NonFatal(ex) => 
+                        promise.tryFailure(ex)
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams/Implementation/Sinks.cs]: comment can be removed

File: src/core/Akka.Streams.Tests/IO/FileSourceSpec.cs [hunk 4]
```diff
@@ -100,6 +102,55 @@ public void FileSource_should_read_contents_from_a_file()
...
+                    return chunks;
+                });
+
+                sub.Request(5000);
+
+                var expectedChunk = nextChunk();
+                for(int i=0; i<10; ++i)
+                {
+                    var actual = c.ExpectNext().DecodeString(Encoding.UTF8);
+                    actual.Should().Be(expectedChunk);
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams.Tests/IO/FileSourceSpec.cs]: why not simply `.Be(nextChunk())`?

File: src/core/Akka.Streams/Implementation/IO/FileSubscriber.cs [hunk 5]
```diff
@@ -76,6 +82,9 @@ protected override void PreStart()
             try
             {
                 _chan = _f.Open(_fileMode, FileAccess.Write);
+                if (_startPosition > 0) {
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams/Implementation/IO/FileSubscriber.cs]: either no curly braces or in the next line, but no java style :stuck_out_tongue:

File: src/core/Akka.Streams/Implementation/IO/FilePublisher.cs [hunk 6]
```diff
@@ -101,6 +108,9 @@ protected override void PreStart()
             try
             {
                 _chan = _f.Open(FileMode.Open, FileAccess.Read);
+                if (_startPosition > 0) {
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams/Implementation/IO/FilePublisher.cs]: either no curly braces or in the next line, but no java style :stuck_out_tongue:

File: src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs (hunk 2 above)
Comment (Arkatufus) [inline src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs]: I'm not sure either, the origin code is a flatMap to ExecutionContexts, so the closest I can think of is to await the Task<Task<IOResult>> TMat. It does work though, I ran all the tests.

**Comment by Aaronontheweb** [comment]
Sorry for the delay on this @Arkatufus - got @Silv3rcircl3 's upstream changes merged in and we're going to work on bringing this in too now. My fault for the delay; was preoccupied getting 1.2.1 out today.

**Comment by Arkatufus** [comment]
It's no problem, really :)

**Pull request merged**